Deprecate ClassMetadataInfo

greg0ire opened the pull request:
The deprecation happened 7 years ago: https://github.com/doctrine/doctrine2/pull/249

Here is how I proceeded:

1. I moved all the code back from `ClassMetadataInfo` to `ClassMetadata`
2. I aliased `ClassMetadataInfo` to `ClassMetadata`.
3. I kept a `ClassMetadataInfo` class definition in an `if (\false)` statement to provide IDE autocompletion.

SenseException committed:
Set copy-method as deprecated

SenseException committed:
Add deprecation of EntityManagerInterface::copy() to upgrade information

greg0ire committed:
Deprecate ClassMetadataInfo

greg0ire committed:
Segregate legacy method test

We would not want to mark the whole test as legacy.

Majkl578 commented on tests/Doctrine/Tests/ORM/Mapping/ClassMetadataInfoTest.php:
```diff
@@ -0,0 +1,32 @@
+     */
+    public function testTheClassIsDeprecated()
+    {
+        class_exists(ClassMetadataInfo::class);
```
Should at least has assertTrue(). Maybe instantiating ReflectionClass would be better as it would throw i.e. once CMI BC is dropped.

greg0ire replied:
Once CMI BC is dropped, the class will no longer exist, and this will throw too. Even without the `assertTrue`, right? I mean there is the `@expectedDeprecation`, and it will no longer be satisfied, I don't know why you want me to add `assertTrue`. I'll add it but I don't understand what the point is.

Majkl578 replied:
Doesn't matter really, just thought it may be easier to see failed assertion first than _no deprecation thrown_ or whatever that message looks like.

Majkl578 commented on tests/Doctrine/Tests/ORM/Mapping/ClassMetadataInfoTest.php:
```diff
@@ -0,0 +1,32 @@
+     * @group legacy
+     * @expectedDeprecation Doctrine\ORM\Mapping\ClassMetadataInfo is deprecated since 2.x and will be removed in 3.0. Use Doctrine\ORM\Mapping\ClassMetadata instead.
+     */
+    public function testTheClassIsDeprecated()
```
return types here and below please

Majkl578 commented on lib/Doctrine/ORM/Mapping/ClassMetadataInfo.php:
```diff
@@ -19,3381 +19,16 @@
-use Doctrine\Common\ClassLoader;
-use Doctrine\ORM\Cache\CacheException;
+@trigger_error(sprintf(
+    '%s is deprecated since 2.x and will be removed in 3.0. Use %s instead.',
```
Note: Revisit message later, it may be ComponentMetadata instead of ClassMetadata.

Majkl578 commented on lib/Doctrine/ORM/Mapping/ClassMetadataInfo.php:
```diff
@@ -19,3381 +19,16 @@
-     * @todo 3.0 Remove this. PersisterHelper should fix it somehow
-     */
-    public function getTypeOfField($fieldName)
-  
```
Why do we need this? Should also use ::class.

greg0ire replied:
We don't: it says "remove this"

Majkl578 replied:
This comment was for:
```php
class_exists('Doctrine\ORM\Mapping\ClassMetadata');
```
GitHub somehow put it on wrong line, maybe too large diff...

Majkl578 commented:
Also please add note to UPGRADE document, thanks.

greg0ire committed:
Add deprecation note to the upgrade guide

Majkl578 committed:
Bump version to 2.7-dev

greg0ire commented:
Triggering the build again

greg0ire committed:
Add listener for every phpunit configuration

greg0ire committed:
Run test in separate process

Otherwise, ClassMetadataInfo gets autoloaded in some other test and the
deprecation is not triggered.

greg0ire committed:
Add explanatory comment

greg0ire commented:
Rebased

lcobucci commented:
@greg0ire the destination branch should be `2.7`

greg0ire commented:
@lcobucci fixed :)

SenseException committed:
Create 2.7 upgrade headline for deprecation changes

Ocramius committed:
Merge pull request #6803 from SenseException/deprecate-copy-method

Deprecation of EntityManager copy method

greg0ire committed:
Move the class aliasing to the deprecated class

It should go when the class finally dies.

greg0ire committed:
Use class constants

Majkl578 commented on lib/Doctrine/ORM/Query/ResultSetMappingBuilder.php:
```diff
@@ -253,12 +253,12 @@ private function getColumnAliasMap($className, $mode, array $customRenameColumns
      * Adds the mappings of the results of native SQL queries to the result set.
      *
-     * @param ClassMetadataInfo $class
+     * @param ClassMetadata $class
```
The comments should be re-aligned everywhere.

Majkl578 reviewed (commented):
LGTM, we just need to revisit the message later and consider adding forward-compatible alias to ComponentMetadata.

greg0ire committed:
Realign comments

greg0ire commented:
@Majkl578 I realigned the comments… with a tool… which I won't mention :trollface:

alcaeus commented on lib/Doctrine/ORM/Mapping/ClassMetadataInfo.php:
```diff
@@ -19,3381 +19,19 @@
-     * @todo 3.0 Remove this. PersisterHelper should fix it somehow
-     */
-    public function getTypeOfField($fieldName)
-  
```
PhpStorm interprets `class_alias`, and I hope other IDEs do as well. There's no need to declare the class like this.

SenseException commented on lib/Doctrine/ORM/Mapping/ClassMetadataInfo.php:
```diff
@@ -19,3381 +19,10 @@
-     * @todo 3.0 Remove this. PersisterHelper should fix it somehow
-     */
-    public function getTypeOfField($fieldName)
-  
```
With `class_alias()`, `ClassMetadataInfo` is shown as unknown class in my PHPStorm IDE. I guess we can live with that since we don't code for IDEs and an error will be shown in the IDE telling us, that something important changed.

SenseException replied:
Hm, it took the deleted side of the diff instead of the one with the class_alias. Sorry.

SenseException reviewed (commented):
You missed one `ClassMetadataInfo` in `lib/Doctrine/ORM/Mapping/Driver/XmlDriver.php`

Commit:
Remove unwanted space chars

Majkl578 commented:
Just realized - we can't use `class_alias` in the file and rely on autoloading - will not work with Composer's authoritative classmap. :(
This would require an explicit autoload in composer.json then:
```json
"autoload": {
    "files": [
        "lib/Doctrine/ORM/Mapping/ClassMetadataInfo.php"
    ]
}

greg0ire committed:
Add extra rationale for dead code block

greg0ire commented:
I just fixed it by restoring the dead code block (I added an extra comment to explain).
How to test I'm not bullshitting?

```
cd /tmp
mkdir whatever
cd whatever
cat > composer.json <<'EOT'
{
    "repositories": [{
        "type": "vcs",
        "url": "https://github.com/greg0ire/doctrine2"

    }],
    "require": {
        "doctrine/orm": "dev-deprecate_classmetadatainfo"
    }
}
EOT
cat > test.php <<'EOT'
<?php

require 'vendor/autoload.php';

use Doctrine\ORM\Mapping\ClassMetadataInfo;

var_dump(class_exists(ClassMetadataInfo::class));
EOT
composer install -a
php test.php
```

greg0ire commented:
Switched to a centralized class deprecation system (@Majkl578 's idea, I like it, do you?)

SenseException commented:
@greg0ire Will it stay in the project or was this created just for `ClassMetadataInfo`? It's currently one affected class, but could be used for coming deprecations, if this was your intention.

greg0ire commented:
I think it will stay in doctrine2 because it's hard to externalize without adding overhead (see [twitter thread](https://twitter.com/greg0ire/status/952188793792139265)), and yeah, the goal is to use it for more classes.

nicolas-grekas commented on composer.json:
```diff
@@ -30,13 +30,15 @@
     "autoload": {
-        "psr-4": { "Doctrine\\ORM\\": "lib/Doctrine/ORM" }
+        "psr-4": { "Doctrine\\ORM\\": "lib/Doctrine/ORM" },
+        "files": ["deprecating_autoloader.php"]
```
the commit introducing this line and the corresponding logic should be removed: it couples Doctrine to Composer, making it not-PSR4 compliant anymore. Furthermore, adding a dedicated autoloaded for Doctrine's deprecation just doens't scale. Imagine if all libs were to do so? No way on my side.

greg0ire committed:
Add deprecated annotation

greg0ire commented:
Moved the controversial autoloading to another PR, so that this is easy to merge.

greg0ire committed:
Include package name in deprecation messages

The clearer the better.

greg0ire commented:
Don't merge, this crashes:

```
<?php

require 'vendor/autoload.php';

use Doctrine\ORM\Mapping\ClassMetadata;
use Doctrine\ORM\Mapping\ClassMetadataInfo;

class A
{
    public function __construct(ClassMetadataInfo $a)
    {
    }
}

new A(new ClassMetadata('whatever'));
```

greg0ire committed:
Ensure old type hints still work

Changing the order of the 2 tests is important because the first one
autoloads CMI

greg0ire commented:
Forgot to say that it is now "fixed", but the fix is a bit ugly

lcobucci commented:
@greg0ire moving this to 2.8.0 as we have conflicts to address and want to release 2.7.0 ASAP.
